### ISSUE TITLE
Hyper hangs with long-running concurrent requests

### ISSUE THREAD

**Issue by cake4289** [hyperium/hyper#1779]
If a computation takes a long time to complete on the server, it can cause hyper to fail to serve other requests. Basically, `tokio::spawn` is not getting called for each request.

```rust
extern crate hyper;

use std::time::Duration;
use std::thread::sleep;

use hyper::{Body, Method, Response};
use hyper::rt::Future;
use hyper::service::service_fn_ok;

fn main() {
    let sockaddr = "127.0.0.1:10022".parse().unwrap();
    let server = hyper::Server::bind(&sockaddr)
        .serve(|| service_fn_ok(|request| {
            println!("Received request");
            match (request.method(), request.uri().path()) {
                (&Method::GET, "/sleep") => {
                    println!("Sleeping...");
                    sleep(Duration::from_secs(20));
                    println!("Sleep completed");
                    Response::new(Body::from(""))
                }
                (&Method::GET, "/") | (&Method::GET, "") => {
                    Response::new(Body::from(r#"
                        <!doctype HTML>
                        <html>
                            <button onclick="
                                var req = new XMLHttpRequest();
                                req.open('GET', '/sleep');
                                req.send(null);
                            ">
                                Sleep
                            </button>
                        </html>
                    "#))
                }
                _ => panic!()
            }
        })).map_err(|e| println!("Server error: {}", e));

    hyper::rt::run(server);
}
```

To replicate the issue, run this code and go to http://localhost:10022 - pressing "Sleep" first and then refreshing the page (or loading it in a new tab) will spuriously fail until the sleep (representing a long calculation) is over. I have not found a canonical way to solve this problem. Doing the `sleep` inside a lazy future does not work. Of course I can spawn a new thread but this seems to defeat the point of using tokio as a runtime. Is there a canonical way to solve this problem? Or are hyper services inherently synchronous?

**Comment by davidbarsky** [comment]
`std::thread::sleep` is a thread-blocking operation, which means the entire reactor + any futures spawned to that reactor would block the reactor thread. I'd suggest using Tokio's [timer](https://docs.rs/tokio-timer/0.2.10/tokio_timer/) to place a _future_—not the runtime!—to sleep. This will allow other futures/requests to make progress in the the meantime.

**Comment by cake4289** [comment]
Hello David, I am only using `sleep` here as an example - in reality I have a long-running calculation and want to serve other requests in the meantime.

Is it really the case that async hyper (using futures) is really synchronous? It is incredibly surprising to me that a long-running calculation (say, a large file or slow database request) could gum up the server when the tokio runtime uses a threadpool.

For reference, here is an async version:

```rust
extern crate futures;
extern crate hyper;

use std::error::Error;
use std::time::Duration;
use std::thread::sleep;

use futures::lazy;
use futures::Poll;
use hyper::{Body, Method, Request, Response};
use hyper::rt::Future;
use hyper::service::Service;

struct MyService;

impl Service for MyService {
    type ReqBody = Body;
    type ResBody = Body;
    type Error = Box<Error + Send + Sync + 'static>;
    type Future = Box<Future<Item=Response<Body>, Error=Self::Error> + Send>;

    fn call(&mut self, request: Request<Body>) -> Self::Future {
        println!("Received request");
        match (request.method(), request.uri().path()) {
            (&Method::GET, "/sleep") => {
                Box::new(lazy(|| {
                    println!("Sleeping...");
                    sleep(Duration::from_secs(20));
                    println!("Sleep completed");
                    Ok(Response::new(Body::from("")))
                }))
            }
            (&Method::GET, "/") | (&Method::GET, "") => {
                Box::new(lazy(|| {
                    Ok(Response::new(Body::from(r#"
                        <!doctype HTML>
                        <html>
                            <button onclick="
                                var req = new XMLHttpRequest();
                                req.open('GET', '/sleep');
                                req.send(null);
                            ">
                                Sleep
                            </button>
                        </html>
                    "#)))
                }))
            }
            _ => panic!()
        }
    }
}

impl Future for MyService {
    type Item = MyService;
    type Error = Box<Error + Send + Sync + 'static>;

    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
        Ok(From::from(MyService))
    }
}

fn main() {
    let sockaddr = "127.0.0.1:10022".parse().unwrap();
    let server = hyper::Server::bind(&sockaddr)
        .serve(|| MyService)
        .map_err(|e| println!("Server error: {}", e));

    hyper::rt::run(server);
}
```

**Comment by sfackler** [comment]
> Is it really the case that async hyper (using futures) is really synchronous?

I think you may be confused about what "asynchronous" means in the context of Tokio. It specifically refers to nonblocking IO - and tasks that cooperatively share a single OS thread by yielding to others when they're unable to make active progress because e.g. they're waiting on data from the network.

The tokio runtime is multi-threaded - there is a thread per CPU core by default. Blocking on a long running operation prevents other requests from running on that thread for the duration. If you have 8 CPU cores and 8 tasks are blocked, no progress can be made.

**Comment by cake4289** [comment]
The reason I am confused is that this problem does *not* occur with tokio on its own. Is it because requests queued on a single connection and pinned to the same thread? Here is a tokio example which does *not* block the other thread - it takes 5 seconds *only*

```rust
extern crate futures;
extern crate tokio;

use std::time::Duration;
use std::thread::sleep;

use futures::lazy;

fn main() {
    tokio::run(lazy(|| {
        for i in 0..4 {
            tokio::spawn(lazy(move || {
                println!("{} Sleeping...", i);
                sleep(Duration::from_secs(5));
                println!("{} Sleep completed", i);
                Ok(())
            }));
        }
        for i in 4..100 {
            tokio::spawn(lazy(move || {
                println!("{} Running", i);
                Ok(())
            }));
        }
        Ok(())
    }));
}
```

Meanwhile my example above blocks the request even with only 2 requests.

**Comment by sfackler** [comment]
I don't remember exactly how tasks are pinned to threads, but it wouldn't be all that surprising if they ended up on the same one.

**Comment by cake4289** [comment]
Thank you very much for clarifying, I guess I can't rely on hyper to distribute requests to different threads on the same connection, and that's why my code is getting stuck. I will look at adding another layer of indirection to my slow-running code (so it can work more like an async request as intended).

**Comment by seanmonstar** [comment]
> hyper to distribute requests to different threads on the same connection

For HTTP/1, hyper doesn't **not** spawn new tasks for each request of the same connection. Even if it did, it will not start processing a new request until the previous response is sent. So spawning a new task would just reduce performance.

In general, you cannot do blocking operations in an async task. You should have some other mechanism to run those operations, like a threadpool.

**Comment by cake4289** [comment]
Thanks Sean, I also found this in the `tokio` docs:

```
Cooperative scheduling is used to schedule tasks on executors. A single executor is expected to manage many tasks across a small set of threads. There will be a far greater number of tasks than threads. There also is no pre-emption. This means that when a task is scheduled to execute, it blocks the current thread until the poll function returns.

Because of this, it is important for implementations of poll to only execute for very short periods of time. For I/O bound applications, this usually happens automatically. However, if a task must run a longer computation, it should defer work to a blocking pool or break up the computation into smaller chunks and yield back to the executor after each chunk.
```

Which was counter to my mental model of how it worked.

I have added a `tokio_threadpool` and used the `spawn_handle` function to run computations (it returns a future). I will code up an example since this could be a common use case (running computations in response to an http request).

**Comment by cake4289** [comment]
Even better -

https://docs.rs/tokio-threadpool/0.1/tokio_threadpool/fn.blocking.html
